### PR TITLE
download squarespace escaped video embeds

### PR DESCRIPTION
**Pull request by galgeek**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Captures videos embedded in Squarespace sites in escaped html, like the youtube embeds in Squarespace sites noted in issue #21294.

This can replace #21802 already initially reviewed by @dstftw.

### CODE REVIEW

File: youtube_dl/extractor/generic.py [hunk 1]
```diff
@@ -2395,6 +2395,9 @@ def _real_extract(self, url):
         # Unescaping the whole page allows to handle those cases in a generic way
         webpage = compat_urllib_parse_unquote(webpage)
 
+        # capture escaped squarespace and other embeds
+        webpage = unescapeHTML(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: Do not unescape the whole page. You must only unescape embedded piece.

**Comment by galgeek** [comment]
Thanks, @dstftw!

I've updated, finding all squarespace video embeds, and escaping before searching for video URL.  

I've checked using these 3 URLs:

http://www.ootboxford.com/
https://www.harvardballetcompany.org/past-productions
http://www.immediategratification.org/

File: youtube_dl/extractor/generic.py [hunk 2]
```diff
@@ -2532,6 +2532,18 @@ def _real_extract(self, url):
...
+        matches = re.findall(r'<div class="[^"]*?sqs-block-video[^>]*>', webpage)
+        if matches:
+            sqs_urls = set()
+            for m in matches:
+                sqs_url = re.search(r'<iframe[^>]+?src=[^"]*"(//[^\\]+)', unescapeHTML(m))
+                if sqs_url:
+                    sqs_urls.add(sqs_url.group(1))
+            if sqs_urls:
+                return self.playlist_from_matches(
+                    sqs_urls, video_id, video_title)
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: One more time: you must unescape and expand embed code **in place** leaving the rest to the generic extractor. You **must not** extract iframes on your own.

Comment (galgeek) [reply youtube_dl/extractor/generic.py]: Thanks, @dstftw!
I've updated, and the new code should be pretty close to what you're looking for.

And I've tested again, using the three problem sites, and all again download with the new code.

**Comment by galgeek** [comment]
Thanks, @dstftw!
I've updated, and the new code should be pretty close to what you're looking for.

**Comment by galgeek** [comment]
I've verified that Issue 21294 is still present in youtube-dl version 2019.08.13, and this PR still fixes the issue.

File: youtube_dl/extractor/generic.py [hunk 3]
```diff
@@ -2395,6 +2395,12 @@ def _real_extract(self, url):
         # Unescaping the whole page allows to handle those cases in a generic way
         webpage = compat_urllib_parse_unquote(webpage)
 
+        # unescape squarespace video embeds
+        sqs_videos = re.findall(r'<div class="[^"]*?sqs-video-wrapper[^>]*>', webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: 1. `re.sub`.
2. Regex must match any quotes kind, must allow other attributes before class.
3. Add a test.

**Comment by galgeek** [comment]
@dstftw thanks!

I've updated, and I've verified that the latest updates resolve #21294

File: youtube_dl/extractor/generic.py [hunk 4]
```diff
@@ -2395,6 +2406,13 @@ def _real_extract(self, url):
         # Unescaping the whole page allows to handle those cases in a generic way
         webpage = compat_urllib_parse_unquote(webpage)
 
+        # unescape re.sub replacement
+        def unescape_resub(m):
+            return unescapeHTML(m.group(0))
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: Just put in lambda, no need to extract it.

File: youtube_dl/extractor/generic.py [hunk 5]
```diff
@@ -2075,6 +2075,17 @@ class GenericIE(InfoExtractor):
             },
             'playlist_count': 6,
         },
+        {
+            # Squarespace video embed, 2019-08-28
+            'url': 'http://ootboxford.com',
+            'info_dict': {
+                'id': 'Tc7b_JGdZfw',
+                'title': 'Out of the Blue, at Childish Things 10',
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: Test does not work.

**Comment by galgeek** [comment]
@dstftw thank you for your help with this!  

It's great to see the fix in master, and I learned a lot, too.

**Pull request closed without merging**